File: src/schema/compound.ts
import { validateTransitions } from '../helpers/stateMachine.js';

import { TypeError } from '../helpers/error.js';

import type { Modifier } from '../types/common.js';
import type { SchemaCompoundType } from '../schema/index.js';

export type SchemaTypeModifierName =
  | 'Array'
  | 'Map'
  | 'Optional'
  | 'OneOf'
  | 'Ref'
  | 'Link';

type NestWithoutReplacement<T extends string, U extends string> = {
  [K in T]: Modifier<K, U | Modifier<Exclude<T, K>, U>>;
}[T];

type OneOfPair<T extends string> = {
  [K in T]: `${K},${Exclude<T, K>}`;
}[T];

type OneOfTriplet<T extends string> = {
  [K in T]: `${K},${OneOfPair<Exclude<T, K>>}`;
}[T];

export type SchemaCompoundTypeExpression<T extends string> =
  | NestWithoutReplacement<'Array' | 'Map' | 'Optional', T>
  | Modifier<'OneOf', OneOfPair<T> | OneOfTriplet<T>>
  | Modifier<'Ref', T>
  | Modifier<'Link', `${string}/${string}`>;

const VALID_TRANSITIONS = {
  '<': ['<', '_'],
  _: ['>', ','],
  ',': ['<', '_'],
  '>': ['>', ',']
};
type TokenState = keyof typeof VALID_TRANSITIONS;

export function parseExpression(label: string, exp: string) {
  type Parsed = Omit<SchemaCompoundType<string>, 'size'>;

  const parsed: Record<string, Parsed> = {};
  const tokenized: [TokenState, string][] = [];
  const pattern = /((Array|Map|Optional|OneOf|Ref|Link)<)|([A-Za-z0-9_/]+)|(,|>)/y;
  let matched: RegExpExecArray | null;
  while ((matched = pattern.exec(exp)) != null) {
    if (matched[1]) {
      tokenized.push(['<', matched[2]]);
    } else if (matched[3]) {
      tokenized.push(['_', matched[3]]);
    } else {
      tokenized.push([matched[4] as TokenState, '']);
    }
  }

  if (!validateExpression(tokenized.map(([action]) => action))) {
    throw new TypeError(`Invalid schema expression: ${exp}`);
  }

  if (tokenized.length === 1) {
    parsed[label] = { type: 'Alias', children: [tokenized[0][1]] };
    return parsed;
  }

  const stack: { label: string; record: Parsed }[] = [];
  let curr: { label: string; record: Parsed } | undefined;
  tokenized.forEach(([action, token]) => {
    if (action === '<') {
      const record: Parsed = {
        type: token as SchemaTypeModifierName,
        children: []
      };
      const next = { label, record };
      label += "'";
      parsed[next.label] = next.record;

      if (curr) stack.push(curr);
      curr = next;
    } else if (action === '_') {
      if (curr) curr.record.children.push(token);
    } else if (action === '>') {
      const top = stack.pop();
      if (top && curr) {
        top.record.children.push(curr.label);
      }
      curr = top;
    }
  });
  return parsed;
}

function validateExpression(stateTransition: TokenState[]) {
  const transitionsValid = validateTransitions(
    [',', ...stateTransition, ','],
    (input, output) => VALID_TRANSITIONS[input].includes(output)
  );
  if (!transitionsValid) return false;

  let level = 0;
  for (const action of stateTransition) {
    if (action === '<') {
      level++;
    } else if (action === '>') {
      level--;
    } else if (action === ',') {
      if (level < 1) return false;
    }
  }
  return level === 0;
}

File: python/buffer_ql/schema/compound.py
import re

from ..helpers.state_machine import validate_transitions

VALID_TRANSITIONS = {
    '<': ['<', '_'],
    '_': ['>', ','],
    ',': ['<', '_'],
    '>': ['>', ',']
}


def parse_expression(label, exp):
    parsed = {}
    tokenized = []
    pattern = re.compile(
        r"((Array|Map|Optional|OneOf|Ref|Link)<)|([A-Za-z0-9_/]+)|(,|>)")
    matched = pattern.match(exp)
    while matched:
        if matched.group(1):
            tokenized.append(('<', matched.group(2)))
        elif matched.group(3):
            tokenized.append(('_', matched.group(3)))
        else:
            tokenized.append((matched.group(4), ''))
        matched = pattern.match(exp, pos=matched.end())

    if not validate_expression([action for action, _ in tokenized]):
        raise TypeError(f"Invalid schema expression: {exp}")

    if len(tokenized) == 1:
        parsed[label] = {'type': 'Alias', 'children': [tokenized[0][1]]}
        return parsed

    stack = []
    curr = None
    for action, token in tokenized:
        if action == '<':
            record = {'type': token, 'children': []}
            next_entry = {'label': label, 'record': record}
            label += f"({token})"
            parsed[next_entry['label']] = next_entry['record']
            if curr:
                stack.append(curr)
            curr = next_entry
        elif action == '_':
            if curr:
                curr['record']['children'].append(token)
        elif action == '>':
            if len(stack) > 0:
                top = stack.pop()
                if curr:
                    top['record']['children'].append(curr['label'])
                curr = top
            else:
                curr = None
    return parsed


def validate_expression(state_transition):
    transitions_valid = validate_transitions(
        [','] + state_transition + [','],
        lambda input, output: output in VALID_TRANSITIONS[input]
    )
    if not transitions_valid:
        return False

    level = 0
    for action in state_transition:
        if action == '<':
            level += 1
        elif action == '>':
            level -= 1
        elif action == ',':
            if level < 1:
                return False

    return level == 0
